Configurable Simple, Semantic Regular Expressions & Output Processing

AnastasiyaSoyka opened the pull request:
## Summary

This PR aims to provide more flexibility to the plugin by allowing the user to provide and use custom regular expressions for processing the output of `git describe`. It also allows the user to provide and use their own processing logic as an interceptor or middleman between the plugin and the `version` property of the host project.

## Notes
- Virtually every class has been refactored in some manner (although the API should remain unchanged).
- The readme file was also updated with examples demonstrating usage of the new features.
- The kdoc documentation was also greatly expanded.

MicahZoltu commented:
Thanks for the PR!  I'll try to check this out and get it merged or reviewed this evening.

MicahZoltu reviewed (changes requested):
I like the premise of your changes, in particular making it more configurable, moving the gradle plugin stuff to the new syntax in `build.gradle`, and the test improvements that allow more complete testing by using the `TemporaryFolder` rule and the tests you added.  The observable delegated properties for configuration are also a clever trick as well that I initially was a bit uncomfortable with, but is quickly growing on me.

Unfortunately, I believe you and I have polar opposite opinions on code style and project organization.  Since in the end I am the one that has to maintain this repo, I want the code in it to match my desired styling/organization.

I believe there are a few options we can take from here, and I'll leave it up to you to decide how to move forward.
1. I can work with you in this PR to get the style to match what I would like.  Due to the scope of the change you have made, I suspect this will end up being a fairly long process, but if you prefer this I don't mind sticking it out (be aware, I am very opinionated on my styling).
2. I can take the concepts you have introduced here along with some of the techniques you have provided and re-implement them in a style that aligns with my desires.  I don't mind taking this path, but I don't want to do this without your permission.
3. This repo is CC0 licensed, so you are free to fork it and publish your own version to plugins.gradle.org, no need for attribution even.

Let me know what route you want to take (or another if you have further ideas) and we can move forward on this!

Thanks again for the PR!

AnastasiyaSoyka commented:
> Unfortunately, I believe you and I have polar opposite opinions on code style and project organization. Since in the end I am the one that has to maintain this repo, I want the code in it to match my desired styling/organization.

I'm absolutely open to working with you to rewrite any code that doesn't match your preferred code style, and please feel free to make any changes to my code you like, including tossing it all out if you hate it or something.

The reason I wrote in my own style when making this PR was:
1. The codebase was so tiny it didn't seem to matter.
2. There was no provided styling or contribution guide, so I assumed you were indifferent on the matter.

> The observable delegated properties for configuration are also a clever trick as well that I initially was a bit uncomfortable with, but is quickly growing on me.

I feel the same way; the [docs](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:getting_input_from_the_build) suggest that an extension class should be a POKO with little to no internal logic. The reason I chose to use observable properties here was that (as I understand it) there are no guarantees on when extensions have values assigned to their properties (beyond that it occurs after plugin application), meaning that state updates cannot be handled within the `GitVersioningPlugin` class itself. Please let me know if I've missed something or there is a better way of doing this.

Commit:
Adds support for configurable processor and toString.

Gives users the ability to provide their own `describeResults -> VersionInfo` function.

Gives users the ability to provide their own `VersionInfo.toString()` function.

MicahZoltu commented:
@Ephemerial Check out https://github.com/Zoltu/Gradle.Plugin.Versioning/tree/ephemerial and let me know if that solves the problem you were trying to solve.  In particular, checkout the last few Functional Tests that I added for examples on how to use it.

If you like it, and it aligns with your goals, feel free to sync your fork up to it and assign yourself as the author since you are the one who did most of the heavy lifting: http://stackoverflow.com/a/36189834/879046  If you need help with the git magic to make that happen let me know (instructions depend on your git tooling preference).  If it doesn't align with what you were thinking let me know and we can discuss.

I haven't written up any Readme changes yet, I wanted to wait until everything was settled.  If you like it, feel free to add some Readme changes.

On master I have also added an `.editorconfig` to help avoid some of the basic code style conflicts as well as implemented some Functional tests (thanks to the pointers from your PR!).  I pushed these through ahead of that branch so the branch should be focused primarily on the functionality changes.

Commit:
Add Missing Imports & Remove Redundant Dependency

Commit:
Add Custom VersionInfo Processor Example To Readme

AnastasiyaSoyka commented on src/main/kotlin/com/zoltu/gradle/plugin/GitVersioning.kt:
```diff
@@ -64,25 +72,59 @@ class GitVersioning : Plugin<Project> {
+		/**
+		 * A custom describe processor that will take in a string as input and return a VersionInfo object as output.
+		 */
+		var customDescribeProcessor: Function<String, VersionInfo?> by observable(defaultCustomDescribeProcessor, { property, old, new -> configurationChangeCallback() })
```
I have some concerns here:

Instantiating a new instance of `VersionInfo` in a `build.gradle` script is very clunky because of Groovy's dynamic typing. For example, Intellij IDEA will be unable to resolve the `com.zoltu.gradle.plugin.VersionInfo` symbol and will display an error, even though the code compiles. It's possible to work around this issue by including the plugin as a dependency, but I think it would be better to ask the user for a regex instead of a factory.

MicahZoltu replied:
I actually started with a regex provider, as you had.  The problem there was that it required the user to have a regex with the right number of captures in the right places.  As I was writing the doc-comment and error handling for it, I realized that it was complex to explain to end users and _really_ prone to errors, which is why I fell-back to the factory mechanism instead.

I wonder if there is some way to get strong typing on the `VersionInfo` object in IntelliJ, ideally with auto-complete?  I wonder if Kotlin gradle files have the same problem or not?

Another option would be to have the factory return an array of Strings?  I don't like this because instead of failing fast, you fail slow and debugging is more difficult but presumably logic could be added in the plugin to give very clear messages here.  This will "solve" the IntelliJ problem, but at a cost.

Another option would be to use a RegEx with named captures rather than ordered captures.  I haven't tried this in Java, is it a thing one can do (my RegEx-foo is weak)?

AnastasiyaSoyka replied:
> I wonder if there is some way to get strong typing on the VersionInfo object in IntelliJ, ideally with auto-complete? I wonder if Kotlin gradle files have the same problem or not?

The easiest workaround is just adding a dependency on the plugin in the usual place; that will allow IDEA to resolve the type and remove the error. We could even do this during plugin application to make the process transparent. But asking the user to add another dependency just for this plugin is a really non-ideal solution.

> Another option would be to use a RegEx with named captures rather than ordered captures. I haven't tried this in Java, is it a thing one can do (my RegEx-foo is weak)?

So is mine :sob:

According to [this](http://stackoverflow.com/a/415635/2286903) answer Java has supported named capture groups since Java 7. I'm not sure how this would make a regex provider more error-resistant, though? It seems like it would be easiest to just kindly suggest to the user that they not specify a custom regex provider if they are not 100% confident in what they are doing.

I understand where you're coming from with regards to how difficult it is to safely write or edit a complex regular expression, but I feel like the negatives of the factory-based approach outweigh the negatives of a regex provider.

MicahZoltu replied:
Named captures would make it so the user's desired string didn't have to match the expected order.  For example, they could have `version-patch.minor.major` as their tagging mechanism.  If we just ask for an arbitrary RegEx, then the user is forced into a regex that matches the expected pattern.

Out of curiosity, what is the use case that you are looking to provide for?  Presumably you created this PR to allow you to solve a particular problem, maybe if I understood that it would help understand the best solution for you?

In hindsight, I think allowing for custom expression/factory is what I should have done in the first place, rather than adding support for 2 more variations on the default.  If we move forward with this change, I'll probably do a major version bump and remove the other two in favor of a single default with customization.

MicahZoltu replied:
> adding a dependency on the plugin in the usual place

What is the "usual place" you are referring to?  You mean in the `buildscript` block?  Is it possible for a plugin to expose a class to the rest of Gradle such that IntelliJ picks it up?  If so, that seems like the best option.

AnastasiyaSoyka replied:
> Out of curiosity, what is the use case that you are looking to provide for? Presumably you created this PR to allow you to solve a particular problem, maybe if I understood that it would help understand the best solution for you?

I originally made this PR because I wanted to use a custom implementation of `VersionInfo.toString()` and I didn't realize until later that I could have just done something like:

``` groovy
version "${ZoltuGitVersioning.versionInfo.major}.${ZoltuGitVersioning.versionInfo.minor}" +
        ".${ZoltuGitVersioning.versionInfo.patch}.${ZoltuGitVersioning.versionInfo.commitCount}"
```

Oh well :grimacing:

> What is the "usual place" you are referring to? You mean in the buildscript block? Is it possible for a plugin to expose a class to the rest of Gradle such that IntelliJ picks it up? If so, that seems like the best option.

``` groovy
repositories {
    maven {
        url 'https://plugins.gradle.org/m2/'
    }
}

dependencies {
    compile group: 'gradle.plugin.com.zoltu.gradle.plugin',
            name: 'git-versioning',
            version: '2.0.34'
}
```

But this does mean that we're building an unnecessary extra dependency, which some users might not like.

MicahZoltu replied:
Hmm, I am surprised that adding a compile time dependency would influence the gradle script auto-completion engine in IntelliJ?  If that is how IntelliJ works then that sounds like a bug to me.  IntelliJ should be using resolved `plugins { ... }`  and `buildscript { ... }` stuff for auto-completion and inline checks.

I'm coming around to the idea of allowing for a named-capture regex.  Maybe we can provide both the factory option and the named capture as extensibility points, and then I can see if I can figure out a "good" way to expose the `VersionInfo` object to gradle, and can also see how it works when you have a Kotlin gradle file.  I'll think on it for a bit and then make a decision.

MicahZoltu replied:
A little more research lead me to find that named captures are not available in Kotlin until 1.1, due to the current tight coupling with JDK 6 which doesn't support named captures: https://youtrack.jetbrains.com/issue/KT-12753

I attempted to switch the describe processor from returning `VersionInfo` to returning a map only to realize that the way things are setup now is to fail fast, so if the describe output doesn't initially start with something that matches one of the existing patterns then the plugin will fail before the user gets a chance to apply configuration.  The correct way to resolve this is to not apply versioning until *after* the gradle file has been parsed.  However, if I make that change then the `VersionInfo` data won't be available until after the gradle file has been parsed, which means you wouldn't be able to do `version "${ZoltuGitVersioning.versionInfo.major}..." in your `build.gradle` file.

I could switch it to silently fail, but that leads to user confusion when their project isn't setup right.  I could continue on with not evaluating the versioning until after `build.gradle` has been evaluated and require users to set whatever they want in a closure, but that leads to `build.gradle` complexity for simple scenarios.

I'm going to think on this some more over-night.  At the least, we can get the change you wanted in, which would allow users to specify a custom `toString()` method.  Unlike your workaround, this would make it so the version printed into the JAR manifest would also match your desired structure, and `./gradlew version` would print the desired version.

AnastasiyaSoyka reviewed (commented):
@MicahZoltu I have synced my fork up to HEAD & added a couple of commits to fix minor issues & add an example to readme.

Commit:
Changes custom describe processor to return an array.

Also fixes some major bugs that caused the plugin to fail to work correctly if the user didn't have a version that could be processed by the built-in versioning.  Also removes the previously baked-in support for semantic versioning and instead opts for users supplying their own versioning function if they want to go outside of the default `v1.2-tag` format.

MicahZoltu commented:
I think I figured out a solution to the problem.  With some lazy init magic the plugin will now process versioning at the earliest of the user trying to access the `VersionInfo` configuration object or the post project evaluation.  The one caveat here is that the user must provide any custom describe processors and custom to string functions *before* accessing the `VersionInfo` object.  While this is unfortunate, I think it is a reasonable cost for the increased configurability.  Check out my branch over at https://github.com/Zoltu/Gradle.Plugin.Versioning/tree/iterating-on-describer and either set your fork to match it and we can get this PR merged or (if I don't hear from you) I can create a new PR from that branch and reference this one.

The readme has not yet been updated with the details of the describe processor and there is still some error handling improvements that could be made for when the user attempts to set the describe processor or tostring methods *after* everything has been fully initialized.  I'm keen to get this merged sooner rather than later though, and those things can be done as iterative improvements later on.  Let me know what your thoughts are about this technique, and you can check out the Functional Tests to see how it looks in Groovy.  I think returning a map is a much better solution than the previously attempted `VersionInfo` object because gradle and its dynamicism.  :'(

Commit:
Update Readme To Reflect Latest Changes

- Removed 'How It Works' section from readme because the exact process used could change if they user specifies a custom processor.
- Rewrote and expanded 'Advanced Usage' section with current examples.
- Minor formatting corrections.

AnastasiyaSoyka commented:
@MicahZoltu I have synced my fork to your branch. In addition, I also rewrote the advanced usage section of the readme. Could you review my changes to make sure that they line up with your expectations?

> I think returning a map is a much better solution than the previously attempted VersionInfo object because gradle and its dynamicism. :'(

I 100% agree. In fact, Groovy's map notation makes the resulting code look rather nice.

By the way, I absolutely love your tests!
